### DISCUSSION TITLE
Error when trying to install dotnet 8 with containerbase

### DISCUSSION THREAD

**Discussion by SaharItzhakWhiteSource** [containerbase/base#1712] [Q&A] [answered]
I'm trying to install dotnet 8 - https://dotnet.microsoft.com/en-us/download/dotnet/8.0
When I run "install-tool dotnet 8.0.100 / 8.0.100-rc.2 / 8.0.0-rc.2" 
I get this error:

root@75e836e9afce:/# install-tool dotnet 8.0.100-rc.2
installing v2 tool dotnet v8.0.100-rc.2
dotnet-install: Attempting to download using primary link https://dotnetcli.azureedge.net/dotnet/Sdk/8.0.100-rc.2/dotnet-sdk-8.0.100-rc.2-linux-x64.tar.gz
curl: (22) The requested URL returned error: 404
dotnet-install: The resource at primary link 'https://dotnetcli.azureedge.net/dotnet/Sdk/8.0.100-rc.2/dotnet-sdk-8.0.100-rc.2-linux-x64.tar.gz' is not available.
dotnet-install: Attempting to download using primary link https://dotnetbuilds.azureedge.net/public/Sdk/8.0.100-rc.2/dotnet-sdk-8.0.100-rc.2-linux-x64.tar.gz
curl: (22) The requested URL returned error: 404
dotnet-install: The resource at primary link 'https://dotnetbuilds.azureedge.net/public/Sdk/8.0.100-rc.2/dotnet-sdk-8.0.100-rc.2-linux-x64.tar.gz' is not available.
dotnet_install: Error: Could not find `.NET Core SDK` with version = 8.0.100-rc.2
dotnet_install: Error: Refer to: https://aka.ms/dotnet-os-lifecycle for information on .NET Core support

**Comment by viceice** [comment]
are you using the latest containerbase version?

**Reply by SaharItzhakWhiteSource** [reply]
@viceice 
I've just built a new image to get the latest version (FROM containerbase/base)
And I still get a download error

[dotnet8-log.txt](https://github.com/containerbase/base/files/13296003/dotnet8-log.txt)

**Reply by viceice** [reply]
please past the relevant log parts here

**Reply by rarkins** [reply]
Here was my reproduction of the problem:

```
❯ docker run --rm -it containerbase/base /bin/bash
root@f66b5e7f0969:/# install-tool dotnet 8.0.0-rc.2
[10:24:43.940] INFO (10): Installing tool dotnet v8.0.0-rc.2...
[10:24:48.916] ERROR (10): download failed
    run: 3
    err: {
      "type": "HTTPError",
      "message": "Response code 404 (Not Found)",
      "stack":
          HTTPError: Response code 404 (Not Found)
              at _Request._onResponseBase (/snapshot/dist/containerbase-cli.js:42343:25)
              at _Request._onResponse (/snapshot/dist/containerbase-cli.js:42398:18)
              at ClientRequest.<anonymous> (/snapshot/dist/containerbase-cli.js:42413:17)
              at Object.onceWrapper (node:events:652:26)
              at ClientRequest.emit (node:events:549:35)
              at ClientRequest.emit (node:domain:482:12)
              at HTTPParser.parserOnIncomingClient (node:_http_client:642:27)
              at HTTPParser.parserOnHeadersComplete (node:_http_common:117:17)
              at TLSSocket.socketOnData (node:_http_client:505:22)
              at TLSSocket.emit (node:events:537:28)
      "code": "ERR_NON_2XX_3XX_RESPONSE",
      "timings": {
        "start": 1699439088669,
        "socket": 1699439088671,
        "lookup": 1699439088675,
        "connect": 1699439088685,
        "secureConnect": 1699439088724,
        "upload": 1699439088725,
        "response": 1699439088912,
        "end": 1699439088914,
        "phases": {
          "wait": 2,
          "dns": 4,
          "tcp": 10,
          "tls": 39,
          "request": 1,
          "firstByte": 187,
          "download": 2,
          "total": 245
        }
      },
      "name": "HTTPError",
      "options": {
        "agent": {},
        "decompress": true,
        "timeout": {},
        "prefixUrl": "",
        "ignoreInvalidCookies": false,
        "context": {},
        "hooks": {
          "init": [],
          "beforeRequest": [],
          "beforeError": [],
          "beforeRedirect": [],
          "beforeRetry": [],
          "afterResponse": []
        },
        "followRedirect": true,
        "maxRedirects": 10,
        "throwHttpErrors": true,
        "username": "",
        "password": "",
        "http2": false,
        "allowGetBody": false,
        "headers": {
          "user-agent": "got (https://github.com/sindresorhus/got)",
          "accept-encoding": "gzip, deflate, br"
        },
        "methodRewriting": false,
        "retry": {
          "limit": 2,
          "methods": [
            "GET",
            "PUT",
            "HEAD",
            "DELETE",
            "OPTIONS",
            "TRACE"
          ],
          "statusCodes": [
            408,
            413,
            429,
            500,
            502,
            503,
            504,
            521,
            522,
            524
          ],
          "errorCodes": [
            "ETIMEDOUT",
            "ECONNRESET",
            "EADDRINUSE",
            "ECONNREFUSED",
            "EPIPE",
            "ENOTFOUND",
            "ENETUNREACH",
            "EAI_AGAIN"
          ],
          "backoffLimit": null,
          "noise": 100
        },
        "method": "GET",
        "cacheOptions": {},
        "https": {},
        "resolveBodyOnly": false,
        "isStream": true,
        "responseType": "text",
        "url": "https://dotnetcli.azureedge.net/dotnet/Sdk/8.0.0-rc.2/dotnet-sdk-8.0.0-rc.2-linux-arm64.tar.gz",
        "pagination": {
          "countLimit": null,
          "backoff": 0,
          "requestLimit": 10000,
          "stackAllItems": false
        },
        "setHost": true,
        "enableUnixSockets": false
      }
    }
[10:24:48.925] FATAL (10): download failed
    err: {
      "type": "Error",
      "message": "download failed",
      "stack":
          Error: download failed
              at HttpService.download (/snapshot/dist/containerbase-cli.js:45894:11)
              at async InstallDotnetService.install (/snapshot/dist/containerbase-cli.js:46388:18)
              at async InstallToolService.execute (/snapshot/dist/containerbase-cli.js:51382:11)
              at async InstallToolShortCommand.execute (/snapshot/dist/containerbase-cli.js:51708:14)
              at async InstallToolShortCommand.validateAndExecute (/snapshot/dist/containerbase-cli.js:1344:26)
              at async _Cli.run (/snapshot/dist/containerbase-cli.js:2457:22)
              at async _Cli.runExit (/snapshot/dist/containerbase-cli.js:2465:28)
              at async main (/snapshot/dist/containerbase-cli.js:51851:3)
    }
[10:24:49.096] INFO (10): Installed tool dotnet with errors in 5.1s.
```

**Comment by rarkins** [comment]
@viceice is there any easy way to iterate over the list of available versions?

When I navigate through the dotnet websites I eventually get to https://download.visualstudio.microsoft.com/download/pr/0d7fb51b-f30e-4b84-b4c2-b739ec8f7978/68e9fc71fb2c4f01a9c7f25672caf1d1/aspnetcore-runtime-8.0.0-rc.2.23480.2-linux-x64.tar.gz
 as the linux download URL, and part of that URL may be non-guessable

**Reply by rarkins** [reply]
Related: https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script

**Reply by viceice** [reply]
probably only the dotnet-version datasource.

this is the current download code for dotnet:
https://github.com/containerbase/base/blob/main/src/cli/tools/dotnet/index.ts#L100-L102

**Reply by rarkins** [reply]
`8.0.0-rc.2` is listed there: https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json

But `install-tool dotnet 8.0.0-rc.2` fails, specifically the URL https://dotnetcli.azureedge.net/dotnet/Sdk/8.0.0-rc.2/dotnet-sdk-8.0.0-rc.2-linux-arm64.tar.gz

**Reply by rarkins** [reply]
According to https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/8.0/releases.json the URLs are like https://download.visualstudio.microsoft.com/download/pr/4fb768da-b0ff-495f-8673-869c2f7e59bf/527c1e1d55a9fbcdd86018bc62b07a03/dotnet-runtime-8.0.0-rc.2.23479.6-linux-x64.tar.gz

Is it possible that they changed the download URLs for newer versions?

**Reply by viceice** [reply]
maybe. maybe only for pre-releases 🤷‍♂️

**Reply by rarkins** [reply]
It seems all major versions are available from download.visualstudio.microsoft.com too

**Reply by rarkins** [reply]
Relevant: https://github.com/actions/setup-dotnet/blob/main/src/installer.ts

**Reply by viceice** [reply]
we don't use the installer shell script for good reasons

**Reply by rarkins** [reply]
Yes but we need to understand why 8.0.0-rc.2 doesn't work with our current approach

**Reply by viceice** [reply]
I'll try to debug next week 🤷‍♂️

**Comment by viceice** [comment, marked as answer]
it should now work, the release is available on the old urls, see PR https://github.com/containerbase/base/pull/1733

@rarkins should we support pre-releases? It would make the installer much more complex because we need to fetch the json file to find the download urls.

**Reply by rarkins** [reply]
Ideally we support pre releases if users are likely to want to use them

**Reply by viceice** [reply]
- https://github.com/containerbase/base/issues/1997
- https://github.com/containerbase/base/issues/1998
